### PR TITLE
Add completion callbacks to SANOs

### PR DESCRIPTION
**Pull request by chrsmith**
⚠️ This is part of a stacked PR set, to be merged into `feature/worker-callbacks`. This will not go directly into `main`, until the overall feature is code complete.

This set of commits were earlier proposed as PR https://github.com/temporalio/temporal/pull/11415, but have since been rebased on top of other changes.

---

## What changed?

The API surface area for the worker callbacks feature adds a new `completion_callbacks` field to SANOs, allowing users to register callbacks to be executed when the SANO completes. (Just like standalone Activities.)

This PR implements that feature for SANO, allowing completion callbacks to be added, executed, and for their status to be reported by the `DescribeNexusOperationExecution` endpoint.

Another couple of features were added as well, to maintain parity with standalone Activities as well. SANOs now provide an `on_conflict_policy` and `links`. The behavior of the conflict policy has been patterned closely after how it is implemented for standalone activities. (To avoid duplication for "Link validation", the functionality was moved into the `chasm` package so it could be reused for both SAA and SANO.)

Completion callbacks for SANOs is governed by two configuration options (following the pattern introduced in the previous PR in the stack): `EnableCallbacks bool` which controls if callbacks can be used at all, and a separate `EnabledCallbackKinds callback.EnabledCallbackKinds` defining which forms of callbacks can be accepted. This defaults to just Nexus-callbacks, and will change when Worker-variant callbacks are actually implemented in a subsequent PR.

## Why?

This is part of the overall worker callbacks feature.

## Notes

Something that this implementation does that standalone Activities does not, is provide a stable response for the `Describe-` endpoint.

If you call `DescribeActivityExecution` multiple times when it has completion callbacks, the order of the returned `CallbackInfo`s will be non-deterministic. (Because it loops over a `map` which isn't stable.)

In the SANO case, we first sort the callbacks by their request ID. So you would get the same result for each call to `Describe-`. (But note that we don't make any attempt to preserve the _ordering_ of callbacks. Just that they are are returned in the same order.)

## How did you test it?

- [x] built
- [x] run locally and tested manually
- [x] covered by existing tests
- [x] added new unit test(s)
- [x] added new functional test(s)

## Potential risks

It's essentially a new feature, and so there are lots of edge cases we need to consider. For example, that repeated calls to `Start-` are idempotent. That the `on_conflict_options` are honored, etc. Otherwise, we run the risk of SANO being slightly inconsistent with how completion callbacks work for standalone Activities.

### CODE REVIEW

File: chasm/lib/nexusoperation/operation_statemachine.go [hunk 1]
```diff
@@ -259,7 +259,6 @@ var TransitionTerminated = chasm.NewTransition(
 		nexusoperationpb.OPERATION_STATUS_SCHEDULED,
 		nexusoperationpb.OPERATION_STATUS_STARTED,
 		nexusoperationpb.OPERATION_STATUS_BACKING_OFF,
-		nexusoperationpb.OPERATION_STATUS_CANCELED,
```
Comment (chrsmith) [inline chasm/lib/nexusoperation/operation_statemachine.go]: This is an unrelated bug that I Claude noticed. Because of this line, we would allow you to Terminate an already Canceled SANO. That causes problems because it would "rewrite the failure reason", as well as be confusing.

Canceled is already a terminal state, and so we shouldn't allow Canceled SANOs to be terminated.

File: tests/nexus_test_base.go [hunk 2]
```diff
@@ -34,6 +37,39 @@ func newNexusTestEnv(t *testing.T, useTemporalFailures bool, opts ...testcore.Te
...
+	req.Namespace = cmp.Or(req.Namespace, env.Namespace().String())
+	req.Service = cmp.Or(req.Service, "test-service")
+	req.Operation = cmp.Or(req.Operation, "test-operation")
+	req.RequestId = cmp.Or(req.RequestId, env.Tv().RequestID())
+	if req.ScheduleToCloseTimeout == nil {
+		req.ScheduleToCloseTimeout = durationpb.New(10 * time.Minute)
+	}
+
+	return env.FrontendClient().StartNexusOperationExecution(ctx, req)
+}
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply tests/nexus_test_base.go:55]: I'll remove the version from `nexus_standalone_test.go`. While working on the code I had a larger refactoring for `nexus_standalone_test.go` and `nexus_test_base.go`, but dropped it in the name of keeping the diff more reasonable.

But in this case, it's just duplication that can easily be removed.

[the remark this answers is not in the document]
Comment (chrsmith) [reply tests/nexus_test_base.go:55]: Err no, let me take that back. Updating all the callsites in `nexus_standalone_test.go` would be a bit of churn that would be better suited in a separate PR.

If anyone feels strongly about it, I can send out a PR to `main` today. Otherwise I'll defer it until worker callbacks initial implementation is all checked in.

File: chasm/lib/nexusoperation/link_validator.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"go.temporal.io/server/chasm"
+	"go.temporal.io/server/common/dynamicconfig"
+)
+
+// linkValidator validates links attached to standalone Nexus operation executions.
+type linkValidator struct {
+	// Wrapped so that nexusoperation.linkValidator is a distinct typed.
+	*chasm.LinkValidator
+}
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply chasm/lib/nexusoperation/link_validator.go:13]: I'll fix the typo, but would argue that having the long doc comment doesn't really help with readability. (But happy to add it if anyone disagrees.)

File: chasm/lib/nexusoperation/operation_statemachine.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrsmith) [reply chasm/lib/nexusoperation/operation_statemachine.go]: Will do.

File: chasm/lib/nexusoperation/validator.go [hunk 4]
```diff
@@ -87,6 +98,23 @@ func (v *validator) validateAndNormalizeStartRequest(req *workflowservice.StartN
...
+			return serviceerror.NewInvalidArgument("completion callbacks are not enabled for this namespace")
+		}
+		opts := callback.ValidateOptions{
+			EnabledKinds: v.config.EnabledCallbackKinds(ns),
+		}
+		if err := v.callbackValidator.Validate(ctx, ns, cbs, opts); err != nil {
+			return err
+		}
+	}
+	if err := v.linkValidator.ValidateRequest(ns, req.GetLinks()); err != nil {
```
[the remark this answers is not in the document]
Comment (chrsmith) [reply chasm/lib/nexusoperation/validator.go]: Good catch. Will fix.

File: chasm/lib/nexusoperation/operation.go [hunk 5]
```diff
@@ -405,6 +439,212 @@ func (o *Operation) getOrCreateOutcome(ctx chasm.MutableContext) *nexusoperation
...
+	requestLinks := ctx.Links(o)
+	if len(requestLinks) == 0 {
+		return o.Links
+	}
+	if len(o.Links) == 0 {
+		return requestLinks
+	}
+	all := make([]*commonpb.Link, 0, len(requestLinks)+len(o.Links))
+	all = append(all, requestLinks...)
+	return append(all, o.Links...)
```
Comment (S15) [inline chasm/lib/nexusoperation/operation.go]: This returns either a reference to an existing slice or a new slice if they need to be combined, which is maybe fine (I don't think any callers currently care), but could be a little weird.

File: chasm/links.go
Comment (bergundy) [inline chasm/links.go:1]: Seems more appropriate to put this in go.temporal.io/server/common/links because it's not part of core CHASM

File: service/frontend/service.go
Comment (bergundy) [inline service/frontend/service.go]: I commented on this in the previous PR but I would collapse these two configs.

**Review by bergundy — APPROVED** [review]
Approved with comments.

File: chasm/lib/nexusoperation/operation.go (hunk 5 above)
Comment (chrsmith) [reply chasm/lib/nexusoperation/operation.go]: I don't think that would be something to worry about, since slices are pass by value and easily copied, etc.

Comment (chrsmith) [reply chasm/lib/nexusoperation/operation.go]: Actually, I didn't quite grok the comment at first. Yeah, we shouldn't be returning o.Links on line 543, since that the returned link would alias o.Links and so any mutations would be a problem.

Will fix.

File: service/frontend/service.go
Comment (chrsmith) [reply service/frontend/service.go]: Yup. Will fix.

**Review by S15 — APPROVED** [review]
LGTM (with a polite note on failed checks)
